src/backend/distributed/master/master_citus_tools.c

```diff
@@ -0,0 +1,538 @@
+
+
+/*
+ * StoreErrorMessage gets the error message from connection and stores it
+ * in queryResultString. It should be called only when error is present
+ * otherwise it would return a default error message.
+ */
+static void
+StoreErrorMessage(PGconn *connection, StringInfo queryResultString)
+{
+	char *errorMessage = PQerrorMessage(connection);
+	char *firstNewlineIndex = strchr(errorMessage, '\n');
+
+	/* trim the error message at the line break */
+	if (firstNewlineIndex != NULL)
+	{
+		*firstNewlineIndex = '\0';
+	}
+
+	/* put a default error message if no error message is reported */
+	if (errorMessage == NULL)
+	{
+		errorMessage = "An error occurred while running the query";
+	}
+
+	appendStringInfo(queryResultString, "%s", errorMessage);
+}
+
+
+/*
+ * ExecuteCommandsAndStoreResults connects to each node specified in
+ * nodeNameArray and nodePortArray, and executes command in commandStringArray
+ * in sequential order. Execution success status and result is reported for
+ * each command in statusArray and resultStringArray. Each array contains
+ * commandCount items.
+ */
+static void
+ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray, int *nodePortArray,
+							   StringInfo *commandStringArray, bool *statusArray,
+							   StringInfo *resultStringArray, int commmandCount)
+{
+	int commandIndex = 0;
+	for (commandIndex = 0; commandIndex < commmandCount; commandIndex++)
+	{
+		char *nodeName = nodeNameArray[commandIndex]->data;
+		int32 nodePort = nodePortArray[commandIndex];
+		bool success = false;
+		char *queryString = commandStringArray[commandIndex]->data;
+		StringInfo queryResultString = resultStringArray[commandIndex];
+
+		success = ExecuteRemoteQueryOrCommand(nodeName, nodePort, queryString,
+											  queryResultString);
+
+		statusArray[commandIndex] = success;
+	}
+}
+
+
+/*
+ * ExecuteRemoteQueryOrCommand executes a query at specified remote node using
+ * the calling user's credentials. The function returns the query status
+ * (success/failure), and query result. The query is expected to return a single
+ * target containing zero or one rows.
+ */
+static bool
+ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort, char *queryString,
+							StringInfo queryResultString)
+{
+	char *nodeUser = CurrentUserName();
+	PGconn *nodeConnection = ConnectToNode(nodeName, nodePort, nodeUser);
+	bool success = false;
+
+	if (nodeConnection == NULL)
+	{
+		appendStringInfo(queryResultString, "failed to connect to %s:%d", nodeName,
+						 (int) nodePort);
+		return false;
+	}
+
+	PG_TRY();
+	{
+		PGresult *queryResult = PQexec(nodeConnection, queryString);
+		success = EvaluateQueryResult(nodeConnection, queryResult, queryResultString);
+
+		PQclear(queryResult);
+
+		/* close the connection */
+		PQfinish(nodeConnection);
+		nodeConnection = NULL;
+	}
+	PG_CATCH();
+	{
+		StoreErrorMessage(nodeConnection, queryResultString);
+
+		/* close the connection */
+		PQfinish(nodeConnection);
+		nodeConnection = NULL;
```

Comment:
How come we're catching errors here, but not in the parallel case?  Either do it in both, or in neither case.  I'd very slightly favor just doing nothing, for simplicity.  We'll soon have infrastructure handling this.